### PR TITLE
Sema: Fix operator function witness lookup for local adoptees

### LINKED ISSUE #45682: [SR-3092] Function-level nested types cannot conform to Equatable
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-3092      |
   |Radar             | rdar://problem/44479781         |
   |Original Reporter | ellneal (JIRA User)      |
   |Type              | Bug    |
   |Status            | Reopened        |
   |Resolution        |     |
        
<details>
  <summary>Environment</summary>

Swift build number: Xcode 8.1 (8B62) embedded version  
Target OS: iOS  
Build OS: macOS Sierra 16A323  
Build Tool: Xcode Version 8.1 (8B62)

</details>

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 5         |
   |Component/s       | Compiler    |
   |Labels            | Bug        |
   |Assignee          | None      |
   |Priority          | Medium      |

   

   md5: f04b40d7c56cb16b9b79a99c8f79b080

  </details>

**is duplicated by**:  
* [SR-3875](https://bugs.swift.org/browse/SR-3875) Can't conform to Equatable when type is inside a function  
* [SR-5939](https://bugs.swift.org/browse/SR-5939) Type nested in function cannot satisfy operator protocol requirement  
* [SR-6526](https://bugs.swift.org/browse/SR-6526) Can't conform a type declared inside a function to Equatable  
* [SR-8696](https://bugs.swift.org/browse/SR-8696) Types nested in functions and their operator functions.   
* [SR-9956](https://bugs.swift.org/browse/SR-9956) classes defined inside functions can't be made to conform to Equatable?  
* [SR-11482](https://bugs.swift.org/browse/SR-11482) Some protocols cannot be conformed by code-block-local types  
* [SR-13398](https://bugs.swift.org/browse/SR-13398) Nested Equatable synthesis cannot be overidden.  

**relates to**:  
* [SR-14795](https://bugs.swift.org/browse/SR-14795) Unable to conform to Comparable protocol when nested  
* [SR-8696](https://bugs.swift.org/browse/SR-8696) Types nested in functions and their operator functions.   

**Issue Description:**

The compiler fails to compile a struct that conforms to \`Equatable\` inside a function.

``` java
struct RootLevelStruct: Equatable { // compiles fine
    static func ==(lhs: RootLevelStruct, rhs: RootLevelStruct) -> Bool {
        return false
    }
}

class MyClass {
    struct ClassLevelStruct: Equatable { // compiles fine
        static func ==(lhs: ClassLevelStruct, rhs: ClassLevelStruct) -> Bool {
            return false
        }
    }
    
    func someFunction() {
        struct FunctionLevelStruct: Equatable { // Error: Protocol 'Equatable' can only be used a generic constraint because it has Self or associated type requirements
            static func ==(lhs: FunctionLevelStruct, rhs: FunctionLevelStruct) -> Bool {
                return false
            }
        }
    }
}
```

### LINKED ISSUE #59016: Nested type conformance problem
With swift-driver version: 1.45.2 Apple Swift version 5.6.1 (swiftlang-5.6.0.323.66 clang-1316.0.20.12)
Target: arm64-apple-macosx12.0

```swift
enum S0: Int, Comparable { // compiles fine
  static func < (lhs: Self, rhs: Self) -> Bool { lhs.rawValue < rhs.rawValue }
  case A, B
}

struct Test {
  func test() {
    enum S1: Int, Comparable { // fails
      static func < (lhs: Self, rhs: Self) -> Bool { lhs.rawValue < rhs.rawValue }
      case A, B
    }
  }
}
```

should compile, but yields:

```
x.swift:8:10: error: type 'S1' does not conform to protocol 'Comparable'
    enum S1: Int, Comparable {
         ^
x.swift:8:14: note: enum declares raw type 'Int', preventing synthesized conformance of 'S1' to 'Comparable'
    enum S1: Int, Comparable {
             ^
Swift.FloatingPoint:3:24: note: candidate would match if 'S1' conformed to 'FloatingPoint'
    public static func < (lhs: Self, rhs: Self) -> Bool
                       ^
Swift.BinaryInteger:4:24: note: candidate has non-matching type '<Self, Other> (Self, Other) -> Bool'
    public static func < <Other>(lhs: Self, rhs: Other) -> Bool where Other : BinaryInteger
                       ^
Swift._Pointer:2:24: note: candidate would match if 'S1' conformed to '_Pointer'
    public static func < (lhs: Self, rhs: Self) -> Bool
                       ^
Swift.Strideable:2:35: note: candidate has non-matching type '<Self> (Self, Self) -> Bool'
    @inlinable public static func < (x: Self, y: Self) -> Bool
                                  ^
Swift.StringProtocol:4:35: note: candidate would match if 'S1' conformed to 'StringProtocol'
    @inlinable public static func < <RHS>(lhs: Self, rhs: RHS) -> Bool where RHS : StringProtocol
                                  ^
Swift.Comparable:2:17: note: protocol requires function '<' with type '(S1, S1) -> Bool'
    static func < (lhs: Self, rhs: Self) -> Bool
                ^
```

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
Operator function witnesses are looked up globally. However, an additional, qualified lookup is warranted if the conforming type is declared in a local context.

This is technically a source-breaking change, since it may alter the behaviour of code that relies on conformances like

```swift
func foo() {
  enum E: Int, Equatable {
    case a, b

    static func == (lhs: E, rhs: E) -> Bool { ... }
  }
}
```

where the provided `==` implementation is now preferred over the Standard Library global [overload](https://developer.apple.com/documentation/swift/2884181).

Resolves #45682

### CODE REVIEW

**Review by slavapestov — APPROVED** [review]
Nice. I found an older bug report and duped the one you referenced in your commit message: https://github.com/apple/swift/issues/45682. No doubt there are others.

**Comment by AnthonyLatsis** [comment]
@swift-ci please test source compatibility

**Comment by AnthonyLatsis** [comment]
UPASS unrelated

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckProtocol.cpp [hunk 1]
```diff
@@ -1315,25 +1315,41 @@ WitnessChecker::lookupValueWitnesses(ValueDecl *req, bool *ignoringNames) {
...
     }
-  } else {
-    // Variable/function/subscript requirements.
+  }
+
+  if (doQualifiedLookup) {
     auto *nominal = Adoptee->getAnyNominal();
     nominal->synthesizeSemanticMembersIfNeeded(reqName.getFullName());
 
+    // If performed, unqualified lookup would have already found candidates from
```
[the remark this answers is not in the document]
Comment (AnthonyLatsis) [reply lib/Sema/TypeCheckProtocol.cpp]: @slavapestov Would be happy to address this in a follow-up if still relevant.

**Pull request merged**